File: 02_data_prep/ScalaApp/anonymize/src/main/scala/Anonymize.scala
//import org.apache.spark.sql._
//spark.sql("show databases").show()
//spark.sql("use raw")
//spark.sql("show tables").show()
//spark.sql("select * from project1_lz limit 10").show()

//import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._


object Anonymize {
  //Logger.getLogger("org").setLevel(Level.WARN)
  //Logger.getLogger("akka").setLevel(Level.WARN)
  def main(args: Array[String]) {
      val spark = SparkSession.builder.appName("anonymize").getOrCreate()
      import spark.implicits._
      
      val input_s3_uri = args(0)      //something like "s3://project1-lz/upload/Tweets.csv"
      val file_name = input_s3_uri.split("//*").last.split("\\.")(0)
      val output_s3_uri = args(1)    //something like "s3://project1-lz/raw/"
      
      val df = spark.read.option("header", true).csv(input_s3_uri)
      
      //replace each tweeters name with crc bigint
      val dfAnnocrc =  df.withColumn("annonym",crc32($"name")).select("annonym", "tweet_id", "airline", "airline_sentiment","text")
      
      //write back to s3 as parquet
      dfAnnocrc.write.mode("append").parquet(output_s3_uri.concat(file_name).concat("-anon"))
  }
}

//spark-submit --class Anonymize ./Anonymize-assembly-0.1.0.jar s3://project1-lz/upload/Tweets.csv s3://project1-lz/raw/

File: 02_data_prep/anonymize.py
from pyspark.sql import SparkSession
from pyspark import SparkContext
from pyspark.sql.functions import sha2, concat_ws
import sys

def main():
    spark = SparkSession.builder.appName("Anonymize PySpark").getOrCreate()
    args_iter = iter(sys.argv[1:])
    args = dict(zip(args_iter, args_iter))
    
    #sample args for interactive testing
    #args = {'project_bucket': 'project1-lz', 'input_table': 'upload', 'output_table': 'raw', 'database': 'default', 'file_name': 'Tweets.csv'}

    project_bucket = args['project_bucket']
    input_table = args['input_table']
    output_table = args['output_table']
    database = args['database']
    input_s3_uri = 's3://' + project_bucket + '/' + input_table + '/' + args['file_name']
    output_s3_uri = 's3://' + project_bucket + '/' + output_table + '/' + args['file_name'].split('.')[0] + '-anon/'

    # Interactive pyspark from glue development endpoint allows reaading from glue crawlers
    # from awsglue.context import GlueContext
    #glueContext = GlueContext(SparkContext.getOrCreate())

    # Create a dataframe from glue catalog
    #df = glueContext.create_data_frame.from_catalog(database=database, table_name=input_table)

    #Print out information about this data
    #print("Count:  ", df.count())
    #df.printSchema()

    df = spark.read.csv(input_s3_uri, header=True)

    # replace each tweeters name with crc bigint
    dfAnnocrc = df.withColumn("annonym", sha2("name", 256)).select("annonym", "tweet_id", "airline", "airline_sentiment", "text")

    # write back to s3 as parquet
    dfAnnocrc.write.mode("append").parquet(output_s3_uri)



if __name__ == "__main__":
    main()




